### ISSUE TITLE
Questions about SONIC with five-finger dexterous hand

### ISSUE THREAD

**Issue by CaiNiao-rookie** [physical-superintelligence-lab/Psi0#70]
Hi,

We are currently trying to add five-finger dexterous hand control to SONIC, but we have encountered several issues and would appreciate your guidance:

- We are trying to add five-finger dexterous hand control to SONIC, using the SONIC framework with a five-finger dexterous hand for teleoperation to collect data, and fine-tuning the Psi0 model.

- Currently, the SONIC framework can properly control the five-finger dexterous hand, but since the action dimensions of the three-finger and five-finger hands are different, when we use the modified SONIC to collect teleoperation data, the hand data is 0.

- Moreover, considering that the original Psi0 model was trained based on the three-finger dexterous hand, even if data is collected, whether it can be effectively fine-tuned？

- Is it possible to try replacing the dexterous hand with a five-finger one in this project?
-  If it can be tried, which files need to be modified?

Any clarification or recommendations would be greatly appreciated. Thank you!

**Comment by songlin** [comment]
Hi @CaiNiao-rookie, thanks for asking.

We tested fine-tuning Psi-0 using a 78-dim sonic action representation, consisting of a 64-dim sonic latent action and 14-dim hand joints for two Dex3-1 hands. If you are using other dexterous hands, just append the hand joint positions to the 64-dim sonic latent action as we did.

The Psi-0 model works perfectly fine with the SONIC action representation plus any other hand joints. The action header can still be initialized from the pretrained checkpoints even though the checkpoint is trained with other action representations (We used the AMO decoupled whole-body control in the released checkpoints). We even experimented with adding two neck degree-of-freedom without any problem.

As for the code, please refer to [Fine-tuning](https://github.com/physical-superintelligence-lab/Psi0/blob/main/scripts/train/psi0/finetune-real-sonic-psi0.sh), [Data collection](https://github.com/physical-superintelligence-lab/Psi0/blob/main/real/SONIC/README.md) and [Deployment](https://github.com/physical-superintelligence-lab/Psi0/blob/main/real/SONIC/DEPLOYMENT.md).

Best,
Songlin

**Comment by Yewandou7** [comment]
Hi @CaiNiao-rookie, 

For teleoperation, we haven't tested this with a five-finger hand ourselves, but here's how you can try it. 

  1. Publish your hand action into the pose message in [pico_manager_thread_server.py](https://github.com/physical-superintelligence-lab/GR00T-WholeBodyControl/blob/main/gear_sonic/scripts/pico_manager_thread_server.py), under the left_hand_joints / right_hand_joints fields.
  2. Update the hand-joint dimension from 7 to the correct dimensions in [features_sonic_vla.py](https://github.com/physical-superintelligence-lab/GR00T-WholeBodyControl/blob/main/gear_sonic/data/features_sonic_vla.py).

With these, the teleop pipeline should record the five-finger hand data correctly instead of zeros.

Best,
Weiduo

**Comment by CaiNiao-rookie** [comment]
Thanks for rely, I will try！

Best，
Rookie

**Comment by YouXiangyu** [comment]
> Thanks for rely, I will try！
> 
> Best， Rookie

@CaiNiao-rookie Hi, I’m following up on your five-finger dexterous-hand adaptation. Were you eventually able to resolve the data-collection issues and successfully fine-tune or deploy Psi-0?
Could you please share a few details about your setup?
How did you collect the five-finger hand data? Did you use SONIC with PICO teleoperation, and how were the hand joint commands generated and recorded?
Which dexterous hand did you use? Was it the Unitree Dex5?
If it was Dex5, did you represent each hand using all 20 URDF joint positions, or the 16 actuated motor joints? In other words, was your Psi-0 action representation 64 + 20 + 20 = 104 dimensions, or did you use another joint mapping?
Besides modifying pico_manager_thread_server.py and features_sonic_vla.py, were any other changes required for data collection, training, or deployment?
If you have any working configuration, code, or experimental results that you can share, they would be extremely helpful. Thank you!

**Comment by pedrosett** [comment]
Hello guys! Just because it might help, we have collected data and fine tuned psi-0 for the g1 with 5 fingers inspire FTP hands. This is the collection with the datasets that worked. Hope this will help

https://huggingface.co/collections/cloudwalk-research/psi0-cw-rl
